src/PlaybackLooping.cpp

```diff
@@ -0,0 +1,208 @@
+/*!********************************************************************
+*
+ Audacity: A Digital Audio Editor
+
+ PlaybackLooping.cpp
+
+ Anton Gerasimov
+
+ **********************************************************************/
+
+#include "PlaybackLooping.h"
+
+#include "Theme.h"
+#include "AllThemeResources.h"
+
+#include "ZoomInfo.h"
+#include "HitTestResult.h"
+#include "RefreshCode.h"
+#include "TrackPanelMouseEvent.h"
+#include "SelectUtilities.h"
+#include "Track.h"
+
+#include <wx/event.h>
+
+PlaybackLooping::PlaybackLooping()
+{
+	;
+}
+
+PlaybackLooping& PlaybackLooping::GetInstance()
+{
+	static PlaybackLooping instance;
+
+	return instance;
+}
+
+bool PlaybackLooping::IsLoopingEnabled() const
+{
+	return mLoopingEnabled;
+}
+
+void PlaybackLooping::EnableLooping(bool enable)
+{
+	mLoopingEnabled = enable;
+}
+
+bool PlaybackLooping::IsProcessing() const
+{
+	return mProcessingState != ProcessingStates::kIdle;
+}
+
+bool PlaybackLooping::LoopingProcessor(AudacityProject* pProject, wxCoord& mousePosX, ProcessingEvents processingEvents)
+{
+	// This is empty project.
+	if (TrackList::Get(*pProject).size() == 0) // NOTE: need check on record?
+		return false;
+
+	switch (mProcessingState)
+	{
+	case ProcessingStates::kIdle:
+		if (processingEvents == ProcessingEvents::kClick)
+		{
+			const auto& viewInfo = ViewInfo::Get(*pProject);
+			const auto& playRegion = viewInfo.playRegion;
+			mOldPlayRegion = playRegion;
+			//SelectUtilities::LockPlayRegion(*pProject);
+			EnableLooping(true);
+
+			auto loopingPos = viewInfo.PositionToTime(mousePosX, viewInfo.GetLeftOffset(), false);
+			mLoopedPlayRegion.SetStart(loopingPos);
+
+			mProcessingState = ProcessingStates::kEnterLooping;
+		}
+		break;
+
+	case ProcessingStates::kEnterLooping:
+		if (processingEvents == ProcessingEvents::kDrag)
+		{
+			const auto& viewInfo = ViewInfo::Get(*pProject);
+			const auto& playRegion = viewInfo.playRegion;
+
+			if (!IsLoopingEnabled() /*&& !playRegion.Locked()*/)
+				return false;
+
+			auto width = viewInfo.GetTracksUsableWidth();
+			mousePosX = std::max(mousePosX, viewInfo.GetLeftOffset());
+			mousePosX = std::min(mousePosX, viewInfo.GetLeftOffset() + width - 1);
+
+			auto loopingPos = viewInfo.PositionToTime(mousePosX, viewInfo.GetLeftOffset(), false);
+			mLoopedPlayRegion.SetEnd(loopingPos);
+		}
+		break;
+	}
+
+	return true;
+}
+
+/* PlaybackLoopingHandle implementation */
+
+PlaybackLoopingHandle::PlaybackLoopingHandle()
+{
+	;
+}
+
+UIHandle::Result PlaybackLoopingHandle::Click (const TrackPanelMouseEvent& event, AudacityProject* pProject)
+{
+	if (event.event.LeftIsDown())
+	{
+		if (!PlaybackLooping::GetInstance().LoopingProcessor(pProject, event.event.m_x, PlaybackLooping::ProcessingEvents::kClick))
```

Comment:
It is the purpose of a UIHandle object to hold temporary state needed only for the duration of a click-drag-release.  There should not be a static object serving as a state machine.

A UIHandle changes some more permanent object ouside of itself in response to clicks and drags.  That would be the play region.  But I also said the play region should not be a global but per-project.

If I open two Audaicty projects, the new red rectangles do not change independently as they should.